src/libraries/Microsoft.PowerFx.Core/Public/IntellisenseOperations.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.Parser;
+using Microsoft.PowerFx.Syntax;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Intellisense
+{
+    /// <summary>
+    /// Various IntelliSense-like operations.
+    /// </summary>
+    public class IntellisenseOperations
+    {
+        private readonly CheckResult _checkResult;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="IntellisenseOperations"/> class.
+        /// </summary>
+        /// <param name="result"></param>
+        public IntellisenseOperations(CheckResult result)
+        {
+            _checkResult = result;
+        }
+
+        /// <summary>
+        /// Checks whether a call to a function with name <paramref name="functionName" /> is valid with argument list
+        /// <paramref name="args" />. Additionally returns (as an out parameter) the return type of this invocation.
+        /// 
+        /// Note: all arguments must belong to the formula that belongs to this <see cref="CheckResult" />.
+        /// </summary>
+        /// <param name="functionName"></param>
+        /// <param name="args"></param>
+        /// <param name="retType"></param>
+        /// <returns></returns>
+        public bool ValidateInvocation(string functionName, IReadOnlyList<TexlNode> args, out FormulaType retType)
+        {
+            retType = null;
+
+            if (args == null)
+            {
+                throw new ArgumentNullException(nameof(args));
+            }
+
+            foreach ((var arg, var index) in args.Select((value, index) => (value, index)))
+            {
+                if (arg == null)
+                {
+                    throw new ArgumentNullException(nameof(args), $"Argument {index} is null");
+                }
+
+                if (!_checkResult._binding.IsNodeValid(arg))
+                {
+                    throw new ArgumentException($"Argument {index} does not belong to this result");
+                }
+            }
+
+            var types = args.Select(node => _checkResult._binding.GetType(node)).ToArray();
+
+            // TODO: Horrible hack to get function identifier name - how to do this in idiomatic way?

```

Comment:
This is a nasty hack. How to do this properly?

Reply:
Is this just to "parse" the functionName in case it contains a namespace? (essentially TexlParser.cs ParseInvocationWithNamespace )

Reply:
That works on the tokenized stuff, right? One (less hacky way) could be to tokenize the string and then construct an identifier manually, but this would still be hacky. The end goal is to convert dotted string to namespace and name (the opposite of `ToDottedSyntax()` of `DPath`).

Reply:
Can we do this:

Keep line 64 as is, but move to its own dedicated function. 
That "quarantines" the hacky code, and gives us a place to document it and unit test. 

// Parse a function name string into a namespace and name. 
bool TryParseFunctionNameAndNamespace(string functionName, out Identifiier fncIdent)
{
 fncIdent = TexlParser.ParseScript($"{functionName}()").Root.AsCall()?.Head;
return fncIdent != null;
}

And then add a unit test for it. 
"Name" 
"Namespace.Name" 
error

Reply:
Done